shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -698,5 +740,107 @@ class ConstraintSolver extends NormalForms { self: Typer =>
   
-  
+  // a is indexed by i
+  // i is indexed in a
+  // receiver: a, index: i, return type of a[i]
+  def constrainIndex(receiver: SimpleType, index: SimpleType)(implicit ctx: Ctx, raise: Raise): SimpleType 
+        = trace(s"$lvl. Receiver: $receiver, Index: $index") {
+    (receiver.unwrapProxies, index.unwrapProxies) match {
+      // if one of receiver, index is Error, then return error
+      case (_, e @ ClassTag(ErrTypeId, _)) => e
+      case (e @ ClassTag(ErrTypeId, _), _) => e
+      // TupleType: fixed length of array, inherited from ArrayBase
+      case (t @ TupleType(fs), ClassTag(IntLit(value), _)) =>  
+        // check index validity and retrieve corresponding type
+        if (value >= fs.length || value < 0) {
+          err(msg"Out of range!", t.prov.loco)
+        } else {
+          fs(value.toInt)._2.ub
+        }
+      case (t: ArrayBase, ClassTag(Var("int"), _)) =>  
+        t.inner.ub | TypeRef(TypeName("undefined"), Nil)(noProv)
+      case (t: ArrayBase, ClassTag(IntLit(value), _) ) =>
+        if (value < 0) {
+          err(msg"Out of range!", t.prov.loco)
+        } else {
+          t.inner.ub | TypeRef(TypeName("undefined"), Nil)(noProv)
+        }
+      // for the case with type ascription, like k: T, it can be assigned a value or not
+      // this kind of type ascription hides more specific information about the underlying value
+      case (t: ArrayBase, rf: TypeRef) => 
+        val r  = rf.expand
+        constrainIndex(t, r)
+      case (rf: TypeRef, t @ ( ClassTag(IntLit(_), _) |  TypeRef(TypeName("int"), _) | ClassTag(Var("int"), _))) =>
+        val r = rf.expand
+        constrainIndex(r, t)
+      // disallow string indexing like "hello"[0]
+      case s @ ((ClassTag(StrLit(_), _), _) | (TypeRef(TypeName("string"), _), _) | (ClassTag(Var("string"), _), _)) =>
+        err(msg"mlscript doesn't allow string indexing", s._1.prov.loco)
+      // StrLit: literal string; TypeRef: defined but not assigned value, so refer to the type; ClassTag(Var(..)): concat "bruh" "bruh", a variable but not a literal
+      case (_, t @ (
+        ClassTag(StrLit(_), _) | ClassTag(Var("string"), _) |
+        ClassTag(DecLit(_), _) | ClassTag(Var("decimal"), _) |
+        ClassTag(Var("bool"), _) | 
+        ClassTag(Var("true"), _) | ClassTag(Var("false"), _) |
+        FunctionType(_, _) | RecordType(_)
+      )) =>
+        err(msg"The index must be an integer", t.prov.loco)
+      case (_, t: ArrayBase) =>
+        err(msg"The index must be an integer", t.prov.loco)
+      case (t @ (
+        ClassTag(IntLit(_), _) | ClassTag(Var("int"), _) |
+        ClassTag(StrLit(_), _) | ClassTag(Var("string"), _) |
+        ClassTag(DecLit(_), _) | ClassTag(Var("decimal"), _) |
+        ClassTag(Var("bool"), _) | 
+        ClassTag(Var("true"), _) | ClassTag(Var("false"), _) |
+        FunctionType(_, _) | RecordType(_)
+      ), _) =>
+        err(msg"The indexing operation should be acted on an array", t.prov.loco)
```

Comment:
There is no reason to special-cases many possible shapes of types like that! You should simply report an error in the default case at the very end.